### ISSUE TITLE
linter: explicit-module-boundary-types/allowOverloadFunctions do not work as intended

### ISSUE THREAD

**Issue by kuainx** [oxc-project/oxc#20454]
### What version of Oxlint are you using?

1.56.0

### What command did you run?

pnpm oxlint

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "rules": {
    "typescript/explicit-module-boundary-types": [
      "error",
      {
        "allowOverloadFunctions": true
      }
    ]
  }
}
```

### What happened?

```ts
// This function should raise an error, but have no error
export function add(a: number, b: number) {
  return a + b;
}

// This function works well with no error
export function plus(a: number, b: number): string;
export function plus(a: string, b: string): string;
export function plus(a: number | string, b: number | string) {
  return a.toString() + b.toString();
}
```
After enabling the "allowOverloadFunctions" option, it seems that **all** the exported functions no longer need to specify return types.
